openrtb_ext/request_wrapper.go

```diff
@@ -0,0 +1,787 @@
+	if ue.consentDirty {
+		consentJson, err := json.Marshal(ue.consent)
+		if err != nil {
+			return nil, err
+		}
+		if len(consentJson) > 0 {
+			ue.ext["consent"] = json.RawMessage(consentJson)
+		} else {
+			delete(ue.ext, "consent")
+		}
+		ue.consentDirty = false
+	}
+
+	if ue.prebidDirty {
+		prebidJson, err := json.Marshal(ue.prebid)
+		if err != nil {
+			return nil, err
+		}
+		if len(prebidJson) > 0 {
+			ue.ext["prebid"] = json.RawMessage(prebidJson)
+		} else {
+			delete(ue.ext, "prebid")
+		}
+		ue.prebidDirty = false
+	}
+
+	if ue.eidsDirty {
+		if len(*ue.eids) > 0 {
+			eidsJson, err := json.Marshal(ue.eids)
+			if err != nil {
+				return nil, err
+			}
+			ue.ext["eids"] = json.RawMessage(eidsJson)
+		} else {
+			delete(ue.ext, "eids")
+		}
+		ue.eidsDirty = false
+	}
+
+	ue.extDirty = false
+	if len(ue.ext) == 0 {
+		return nil, nil
+	}
+	return json.Marshal(ue.ext)
+
+}
+
+func (ue *UserExt) Dirty() bool {
+	return ue.extDirty || ue.eidsDirty || ue.prebidDirty || ue.consentDirty
+}
+
+func (ue *UserExt) GetExt() map[string]json.RawMessage {
+	ext := make(map[string]json.RawMessage)
+	for k, v := range ue.ext {
+		ext[k] = v
+	}
+	return ext
+}
+
+func (ue *UserExt) SetExt(ext map[string]json.RawMessage) {
+	ue.ext = ext
+	ue.extDirty = true
+}
+
+func (ue *UserExt) GetConsent() *string {
+	if ue.consent == nil {
+		return nil
+	}
+	consent := *ue.consent
+	return &consent
+}
+
+func (ue *UserExt) SetConsent(consent *string) {
+	ue.consent = consent
+	ue.consentDirty = true
+}
+
+func (ue *UserExt) GetPrebid() *ExtUserPrebid {
+	if ue.prebid == nil {
+		return nil
+	}
+	prebid := *ue.prebid
+	return &prebid
+}
+
+func (ue *UserExt) SetPrebid(prebid *ExtUserPrebid) {
+	ue.prebid = prebid
+	ue.prebidDirty = true
+}
+
+func (ue *UserExt) GetEid() *[]ExtUserEid {
+	if ue.eids == nil {
+		return nil
+	}
+	eids := *ue.eids
+	return &eids
+}
+
+func (ue *UserExt) SetEid(eid *[]ExtUserEid) {
+	ue.eids = eid
+	ue.eidsDirty = true
+}
+
+// ---------------------------------------------------------------
+// RequestExt provides an interface for request.ext
+// ---------------------------------------------------------------
+
+type RequestExt struct {
+	ext         map[string]json.RawMessage
+	extDirty    bool
+	prebid      *ExtRequestPrebid
+	prebidDirty bool
+}
+
+func (re *RequestExt) unmarshal(extJson json.RawMessage) error {
+	if len(re.ext) != 0 || re.Dirty() {
+		return nil
+	}
+	re.ext = make(map[string]json.RawMessage)
+	if len(extJson) == 0 {
+		return nil
+	}
+	err := json.Unmarshal(extJson, &re.ext)
+	if err != nil {
+		return err
+	}
+	prebidJson, hasPrebid := re.ext["prebid"]
+	if hasPrebid {
+		re.prebid = &ExtRequestPrebid{}
+		err = json.Unmarshal(prebidJson, re.prebid)
+	}
+
+	return err
+}
+
+func (re *RequestExt) marshal() (json.RawMessage, error) {
+	if re.prebidDirty {
+		prebidJson, err := json.Marshal(re.prebid)
+		if err != nil {
+			return nil, err
+		}
+		if len(prebidJson) > 2 {
```

Comment:
Ah, here is where an empty prebid object is dropped. Do you think it's worth checking the content is only `{}` or empty before we delete it?

Reply:
To be valid json at only 2 characters, I believe it has to be '{}', certainly in this context.